openshift-kube-apiserver/admission/storage/performantsecuritypolicy/admission.go

```diff
@@ -0,0 +1,167 @@
+		len(attributes.GetSubresource()) > 0 {
+		return nil
+	}
+
+	pod, ok := attributes.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attributes, fmt.Errorf("unexpected object: %#v", attributes.GetObject()))
+	}
+
+	ns, err := c.nsLister.Get(pod.Namespace)
+	if err != nil {
+		return fmt.Errorf("error listing pod namespace: %v", err)
+	}
+
+	currentFSGroupChangePolicy := extractCurrentFSGroupChangePolicy(pod)
+	if currentFSGroupChangePolicy == nil {
+		currentFSGroupChangePolicy = getDefaultFSGroupChangePolicy(ns)
+		if currentFSGroupChangePolicy != nil {
+			klog.V(3).Infof("Setting default FSGroupChangePolicy %s for pod %s/%s", *currentFSGroupChangePolicy, pod.Namespace, pod.Name)
+			if pod.Spec.SecurityContext != nil {
+				pod.Spec.SecurityContext.FSGroupChangePolicy = currentFSGroupChangePolicy
+			} else {
+				pod.Spec.SecurityContext = &kapi.PodSecurityContext{
+					FSGroupChangePolicy: currentFSGroupChangePolicy,
+				}
+			}
+		}
+	}
+
+	currentSELinuxChangePolicy := extractCurrentSELinuxChangePolicy(pod)
+	if currentSELinuxChangePolicy == nil {
+		currentSELinuxChangePolicy = getDefaultSELinuxChangePolicy(ns)
+		if currentSELinuxChangePolicy != nil {
+			klog.V(3).Infof("Setting default SELinuxChangePolicy %s for pod %s/%s", *currentSELinuxChangePolicy, pod.Namespace, pod.Name)
+			if pod.Spec.SecurityContext != nil {
+				pod.Spec.SecurityContext.SELinuxChangePolicy = currentSELinuxChangePolicy
+			} else {
+				pod.Spec.SecurityContext = &kapi.PodSecurityContext{
+					SELinuxChangePolicy: currentSELinuxChangePolicy,
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func extractCurrentSELinuxChangePolicy(pod *kapi.Pod) *kapi.PodSELinuxChangePolicy {
+	if pod.Spec.SecurityContext != nil {
+		return pod.Spec.SecurityContext.SELinuxChangePolicy
+	}
+
+	return nil
+}
+
+func extractCurrentFSGroupChangePolicy(pod *kapi.Pod) *kapi.PodFSGroupChangePolicy {
+	if pod.Spec.SecurityContext != nil {
+		return pod.Spec.SecurityContext.FSGroupChangePolicy
+	}
+	return nil
+}
+
+func getDefaultFSGroupChangePolicy(ns *corev1.Namespace) *kapi.PodFSGroupChangePolicy {
+	fsGroupPolicy, ok := ns.Labels[fsGroupChangePolicyLabel]
+	if !ok {
+		return nil
+	}
+	policy := kapi.PodFSGroupChangePolicy(fsGroupPolicy)
+
+	if policy == kapi.FSGroupChangeOnRootMismatch || policy == kapi.FSGroupChangeAlways {
+		return &policy
+	}
+	klog.Warningf("found %s label with invalid fsGroupPolicy: %s", fsGroupChangePolicyLabel, fsGroupPolicy)
+	return nil
+}
+
+func getDefaultSELinuxChangePolicy(ns *corev1.Namespace) *kapi.PodSELinuxChangePolicy {
+	selinuxChangePolicy, ok := ns.Labels[selinuxChangePolicyLabel]
+	if !ok {
+		return nil
+	}
+
+	policy := kapi.PodSELinuxChangePolicy(selinuxChangePolicy)
+
+	if policy == kapi.SELinuxChangePolicyMountOption || policy == kapi.SELinuxChangePolicyRecursive {
+		return &policy
+	}
+
+	klog.Warningf("found %s label with invalid selinuxPolicy: %s", selinuxChangePolicyLabel, selinuxChangePolicy)
```

Comment:
Storing the warning only in the KAS logs won't be helpful. Maybe, in addition, we should also send a warning to the client side so that admins can potentially notice immediately that something is fishy.

Take a look at https://github.com/kubernetes/kubernetes/blob/27e16751538cb316f0c8b8f2384a9968df066baa/plugin/pkg/admission/security/podsecurity/admission.go#L191 as an example from an upstream plugin.

Reply:
same as above.